File: ride-sharing-system/go-without-multithreading/strategy/preferred_vehicle_strategy.go
package strategy
import (
	"app/ride-sharing-system/go-without-multithreading/entities"
	"app/ride-sharing-system/go-without-multithreading/enums"
)

// PreferredVehicleStrategy routes mathematical iterations prioritizing absolute parameter equivalence natively.
type PreferredVehicleStrategy struct {
	PreferredType enums.VehicleType
}

func (s *PreferredVehicleStrategy) SelectRide(availableRides []*entities.Ride) *entities.Ride {
	for _, ride := range availableRides {
		if ride.Vehicle.Type == s.PreferredType {
			return ride // Physically returns the first matched execution node!
		}
	}
	return nil
}

File: ride-sharing-system/java-without-multithreading/strategy/PreferredVehicleStrategy.java
package strategy;
import entities.Ride;
import enums.VehicleType;
import java.util.List;
import java.util.Optional;

public class PreferredVehicleStrategy implements RideSelectionStrategy {
    private VehicleType preferredType;

    public PreferredVehicleStrategy(VehicleType preferredType) {
        this.preferredType = preferredType;
    }

    @Override
    public Optional<Ride> selectRide(List<Ride> availableRides) {
        for (Ride ride : availableRides) {
            if (ride.getVehicle().getType() == preferredType) {
                return Optional.of(ride);
            }
        }
        return Optional.empty(); // Returns safely if mathematically isolated
    }
}
